app/Models/Device.php

```diff
@@ -324,15 +324,33 @@ public function getAttribs()
         return $this->attribs->pluck('attrib_value', 'attrib_type')->toArray();
     }
 
-    public function setLocation($location_text)
+    /**
+     * Update the location to the correct location and update GPS if needed
+     *
+     * @param  \App\Models\Location  $location location data
+     */
+    public function setLocation(Location $location)
     {
-        $location_text = $location_text ? Rewrite::location($location_text) : null;
+        $location->location = $location->location ? Rewrite::location($location->location) : null;
 
-        $this->location_id = null;
-        if ($location_text) {
-            $location = Location::firstOrCreate(['location' => $location_text]);
+        if (! $location->location) {
+            $this->location_id = null;
+
+            return;
+        }
+
+        $coord = array_filter($location->only(['lat', 'lng']));
+        if (! $this->relationLoaded('location') || optional($this->location)->location !== $location->location) {
+            if (! $location->exists) {
```

Comment:
when would it exist? since you create it with `new Location` it never exists?

Reply:
Correct, but people can implement custom location fetching in the OS PHP.  This simple check avoids an sql query in that case.  Could remove it, but I think it is better to have it there than not.